Mathlib/LinearAlgebra/DirectSum/Finsupp.lean

```diff
@@ -12,20 +12,274 @@ import Mathlib.LinearAlgebra.DirectSum.TensorProduct
+    (ι →₀ M) ⊗[R] N ≃ₗ[R] ι →₀ (M ⊗[R] N) :=
+  (congr (finsuppLEquivDirectSum R M ι) (LinearEquiv.refl R N)).trans
+    ((directSumLeft R (fun _ : ι => M) N).trans
+      (finsuppLEquivDirectSum R (M ⊗[R] N) ι).symm)
+
+lemma finsuppLeft_apply_tmul (p : ι →₀ M) (n : N) :
+    finsuppLeft (p ⊗ₜ[R] n) = p.sum (fun i m ↦ Finsupp.single i (m ⊗ₜ[R] n)) := by
+  simp only [finsuppLeft, LinearEquiv.trans_apply, congr_tmul, LinearEquiv.refl_apply]
+  conv_lhs => rw [← Finsupp.sum_single p]
+  rw [LinearEquiv.symm_apply_eq]
+  simp only [map_finsupp_sum]
+  simp only [finsuppLEquivDirectSum_single]
+  rw [← LinearEquiv.eq_symm_apply]
+  simp only [map_finsupp_sum]
+  simp only [directSumLeft_symm_lof_tmul]
+  simp only [Finsupp.sum, sum_tmul]
+
+@[simp]
+lemma finsuppLeft_apply_tmul_apply (p : ι →₀ M) (n : N) (i : ι) :
+    finsuppLeft (p ⊗ₜ[R] n) i = p i ⊗ₜ[R] n := by
+  rw [finsuppLeft_apply_tmul]
+  simp only [Finsupp.sum_apply]
+  conv_rhs => rw [← Finsupp.single_eq_same (a := i) (b := p i ⊗ₜ[R] n)]
+  apply Finsupp.sum_eq_single i
+  · exact fun _ _ ↦ Finsupp.single_eq_of_ne
+  · intro _
+    simp
+
+theorem finsuppLeft_apply (t : (ι →₀ M) ⊗[R] N) (i : ι) :
+    (finsuppLeft t) i = (rTensor N (Finsupp.lapply i)) t := by
+  induction t using TensorProduct.induction_on with
+  | zero => simp
+  | tmul f n => simp only [finsuppLeft_apply_tmul_apply, rTensor_tmul, Finsupp.lapply_apply]
+  | add x y hx hy => simp [LinearEquiv.map_add, hx, hy]
+
+@[simp]
+lemma finsuppLeft_symm_apply_single (i : ι) (m : M) (n : N) :
+    finsuppLeft.symm (Finsupp.single i (m ⊗ₜ[R] n)) =
+      Finsupp.single i m ⊗ₜ[R] n := by
+  simp [finsuppLeft, Finsupp.lsum]
+
+/-- The tensor product of `M` and `ι →₀ N` is linearly equivalent to `ι →₀ M ⊗[R] N` -/
+noncomputable def finsuppRight :
+    M ⊗[R] (ι →₀ N) ≃ₗ[R] ι →₀ (M ⊗[R] N) :=
+  (congr (LinearEquiv.refl R M) (finsuppLEquivDirectSum R N ι)).trans
+    ((directSumRight R M (fun _ : ι => N)).trans
+      (finsuppLEquivDirectSum R (M ⊗[R] N) ι).symm)
+
+lemma finsuppRight_apply_tmul (m : M) (p : ι →₀ N) :
+    finsuppRight (m ⊗ₜ[R] p) = p.sum (fun i n ↦ Finsupp.single i (m ⊗ₜ[R] n)) := by
+  simp [finsuppRight]
+  conv_lhs => rw [← Finsupp.sum_single p]
+  rw [LinearEquiv.symm_apply_eq]
+  simp only [map_finsupp_sum]
+  simp only [finsuppLEquivDirectSum_single]
+  rw [← LinearEquiv.eq_symm_apply]
+  simp only [map_finsupp_sum]
+  simp only [directSumRight_symm_lof_tmul]
+  simp only [Finsupp.sum, tmul_sum]
+
+@[simp]
+lemma finsuppRight_apply_tmul_apply (m : M) (p : ι →₀ N) (i : ι) :
+    finsuppRight (m ⊗ₜ[R] p) i = m ⊗ₜ[R] p i := by
+  rw [finsuppRight_apply_tmul]
+  simp only [Finsupp.sum_apply]
+  conv_rhs => rw [← Finsupp.single_eq_same (a := i) (b := m ⊗ₜ[R] p i)]
+  apply Finsupp.sum_eq_single i
+  · exact fun _ _ ↦ Finsupp.single_eq_of_ne
+  · intro _
+    simp
+
+theorem finsuppRight_apply (t : M ⊗[R] (ι →₀ N)) (i : ι) :
+    (finsuppRight t) i = (lTensor M (Finsupp.lapply i)) t := by
```

Comment:
I'd remove unnecessary parentheses (you did not add parentheses in the previous lemma for example).

Reply:
OK